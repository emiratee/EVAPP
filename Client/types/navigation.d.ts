export interface Screens {
    index: undefined
    history: undefined
    login: undefined
    chatView: {}
    addCredits: undefined
    profile: undefined
    TripInfo: {}
    messages: undefined
    TripCardRedirect: {}
    BookRequest: {}
}

export declare global {
    namespace ReactNavigation {
        interface RootParamList extends Screens { }
    }
}