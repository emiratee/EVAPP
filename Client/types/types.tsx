export type TCar = {
    _id: string,
    model: string,
    color: string,
    seats: number,
    licencePlate: string
}



export type TTrip = {
    _id: string,
    departure: {
        country: string,
        city: string,
        address: string,
        time: string
    },
    destination: {
        country: string,
        city: string,
        address: string,
        time: string
    },
    date: string,
    totalTime: string,
    seats: {
        available: number,
        total: number
    },
    services: {
        smoking: boolean,
        childSeat: boolean,
        pets: boolean,
        alcohol: boolean,
        luggage: boolean,
        comments?: string,
    }
    car: TCar,
    price: string,
    driverID: string,
    passengersIDs: string[],
    successful: boolean

}


export type TUser = {
    _id: string,
    name: string,
    memberSince: string,
    email: string,
    phoneNumber: string,
    password: string,
    cars: TCar[],
    passengerRating: {
        totalReviews: number,
        totalRating: number,
        averageRating: number
    },
    driverRating: {
        totalReviews: number,
        totalRating: number,
        averageRating: number
    },
    tripsAsDriverIDs: string[],
    tripsAsPasangerIDs: string[],
    credits: {
        available: string,
        onHold: string
    } 

}

export type TCarNoId = Omit<TCar, '_id'>
export type TTripNoId = Omit<TTrip, '_id'>
export type TUserNoId = Omit<TUser, '_id'>