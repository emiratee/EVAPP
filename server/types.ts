export type TCar = {
    _id: string,
    model: string,
    color: string,
    seats: number,
    licencePlate: string
}



export type TTrip = {
    _id: string,
    departure: {
        country: string,
        city: string,
        address?: string,
        time: string,
        date: string
    },
    destination: {
        country: string,
        city: string,
        address?: string,
        time: string,
        date: string
    },
    date: string,
    totalTime: string,
    seats: {
        available: number,
        total: number
    },
    services: {
        smoking: boolean,
        childSeat: boolean,
        pets: boolean,
        alcohol: boolean,
        luggage: boolean,
        comments?: string,
    }
    car: TCar,
    price: string,
    driverID: string,
    passengerIDs: TPassengerIDs[],
    successful: boolean
}



export type TPassengerIDs = {
    bookingId: string,
    userId: string,
    name: string,
    status: string,
    seats: number,
    reviewed: boolean
}

export type TUser = {
    _id: string,
    userId: string,
    name: string,
    expoPushToken?: string,
    memberSince: string,
    email: string,
    imageUrl?: string,
    phoneNumber: string,
    password: string,
    cars: TCar[],
    passengerRating: {
        totalReviews: number,
        totalRating: number,
        averageRating: number
    },
    driverRating: {
        totalReviews: number,
        totalRating: number,
        averageRating: number
    },
    tripsAsDriverIDs: string[],
    tripsAsPasangerIDs: string[],
    credits: {
        available: string,
        onHold: string,
        earningsOnHold: string,
    }

}

export type TDeparture = {
    country: string,
    city: string,
    address?: string,
    time?: string,
    date?: string
}

export type TDestination = {
    country: string,
    city: string,
    address?: string,
    time?: string,
    date?: string
}


export type TTrip_search = {
    departure: TDeparture,
    destination?: TDestination,
    date: string,
    seats: number
}

export type TRegisterForm = {
    name: string,
    email: string,
    phoneNumber: string,
    password: string,
    imageUrl?: string,
    expoPushToken?: string
}

export type TImageFormData = {
    append: (name: string, value: any, fileName?: string) => void,
}

export type TCarNoId = Omit<TCar, '_id'>
export type TTripNoId = Omit<TTrip, '_id'>
export type TUserNoId = Omit<TUser, '_id'>