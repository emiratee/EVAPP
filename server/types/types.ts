export type TTrip = {
    _id?: string;
    departure: {
        country: string;
        city: string;
        address: string;
        time: string;
    };
    destination: {
        country: string;
        city: string;
        address: string;
        time: string;
    };
    date: string;
    totalTime: string;
    seats: {
        available: number;
        total: number;
    };
    services: {
        smoking: boolean;
        childSeat: boolean;
        pets: boolean;
        alcohol: boolean;
        luggage: boolean;
        comments?: string;
    };
    car: [];
    price: string;
    driverID: string;
    passengerIDs: {
        bookingId: string;
        userId: string;
        name: string;
        status: string;
        seats: number;
    }[];
    successful: boolean;
    __v?: string;
}