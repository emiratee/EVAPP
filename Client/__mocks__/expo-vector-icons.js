jest.mock('@expo/vector-icons', () => ({

    
    Ionicons: (props) => <View {...props} />, // Mock Ionicons
    FontAwesome: (props) => <View {...props} />, // Mock FontAwesome
  
}))



// const MockedIcons = {
//     Ionicons: (props) => <View {...props} />, // Mock Ionicons
//     FontAwesome: (props) => <View {...props} />, // Mock FontAwesome
//     // Add more mocked icons as needed
// };

