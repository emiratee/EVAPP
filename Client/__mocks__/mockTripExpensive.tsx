import { TTrip } from "../types/types";

const mockTrip: TTrip = {
    _id: '1234567890', // Replace with a unique ID
    departure: {
      country: 'USA',
      city: 'New York',
      address: '123 Main St',
      time: '08:00',
      date: '2023-12-15',
    },
    destination: {
      country: 'USA',
      city: 'Los Angeles',
      address: '456 Elm St',
      time: '16:00',
      date: '2023-12-16',
    },
    date: '2023-12-15',
    totalTime: '8 hours',
    seats: {
      available: 3,
      total: 5,
    },
    services: {
      smoking: false,
      childSeat: true,
      pets: false,
      alcohol: true,
      luggage: true,
      comments: 'No smoking allowed.',
    },
    car: {
      _id: 'car123', // Replace with the car ID
      model: 'Toyota Camry',
      color: 'Silver',
      licencePlate: 'ABC-123',
      seats: 5,
    },
    price: '50000.00', // Price without currency symbol
    driverID: 'driver123', // Replace with the driver's ID
    passengerIDs: [
      {
        bookingId: 'booking1', // Replace with booking ID
        userId: 'user1', // Replace with passenger's ID
        name: 'Alice',
        status: 'Confirmed',
        seats: 1,
      },
      {
        bookingId: 'booking2', // Replace with booking ID
        userId: 'user2', // Replace with passenger's ID
        name: 'Bob',
        status: 'Confirmed',
        seats: 2,
      },
    ],
    successful: true,
  };
  
  export default mockTrip;
  