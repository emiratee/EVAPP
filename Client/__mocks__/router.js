jest.mock('expo-router', () => ({
    useNavigation: () => ({
        navigate: jest.fn(),
    }),
    useFocusEffect: jest.fn(),
    Link: jest.fn(),
    useRouter: () => ({
        router: jest.fn(),
    }),
}));

jest.mock('@react-navigation/native', () => ({
    useNavigation: jest.fn(),
    useFocusEffect: jest.fn(),
}));
