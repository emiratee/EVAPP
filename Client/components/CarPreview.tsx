import { StyleSheet, } from 'react-native'
import { Text, View } from './Themed'
import React, { useEffect, useState } from 'react'
import * as icons from '@expo/vector-icons'
import { TouchableOpacity } from 'react-native'

type Props = {
    item: any,
    setSelectedCar: any,
    selectedCar: any,
    onPress: () => void
}

const CarPreview = (props: Props) => {
    const [isSelected, setIsSelected] = useState<boolean>(props.selectedCar?.id === props.item.id)
    const styles = getDynamicStyles();
    useEffect(() => {
        props.selectedCar?.id === props.item.id ? setIsSelected(true) : setIsSelected(false);
    }, [props.selectedCar])

    return (
        <TouchableOpacity
            style={[styles.wrapper, isSelected && styles.selected]}
            onPress={props.onPress}
        >
            <Text style={{ fontWeight: "bold" }}>{props.item.model}</Text>

            <Text>{props.item.color}</Text>
            <View style={{ flexDirection: 'row' }}>
                <Text>{props.item.seats}</Text>
                <icons.MaterialCommunityIcons name='seat-passenger' size={18} />
            </View>
            <Text>{props.item.licence_plates}</Text>



        </TouchableOpacity>
    )
}

export default CarPreview

const getDynamicStyles = () => {

    return StyleSheet.create({
        wrapper: {
            // backgroundColor: '#eee',
            padding: 10,
            borderRadius: 10,
            borderWidth: 1,
            borderColor: '#a8a8a8',
            height: 100,
            width: 100,
            marginRight: 10,
            justifyContent:'space-between'

        },
        selected: {
            borderWidth: 1,
            borderColor: '#e30000',
        }

    })
}

