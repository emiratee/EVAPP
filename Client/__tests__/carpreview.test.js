import React from 'react';
import { render, fireEvent, getComputedStyle } from '@testing-library/react-native';
import CarPreview from '../components/CarPreview'; 
import * as types from '../types/types';
import renderer from 'react-test-renderer';

import { mockCar1 } from '../__mocks__/mockUserRich'

describe('CarPreview component', () => {
    it('renders correctly with unselected car', () => {
        const setSelectedCar = jest.fn();
        const selectedCar = null;

        const component = renderer.create(
            <CarPreview
                item={mockCar1}
                setSelectedCar={setSelectedCar}
                selectedCar={selectedCar}
                onPress={jest.fn()}
            />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
    it('renders correctly with selected car', () => {
        const setSelectedCar = jest.fn();
        const selectedCar = mockCar1;

        const component = renderer.create(
            <CarPreview
                item={mockCar1}
                setSelectedCar={setSelectedCar}
                selectedCar={selectedCar}
                onPress={jest.fn()}
            />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls onPress when pressed', () => {
        const setSelectedCar = jest.fn();
        const selectedCar = null;
        const onPress = jest.fn();

        const { getByTestId } = render(
            <CarPreview
                item={mockCar1}
                setSelectedCar={setSelectedCar}
                selectedCar={selectedCar}
                onPress={onPress}
            />
        );

        // Simulate a press event
        fireEvent.press(getByTestId('car-preview-wrapper'));

        // Assert that the onPress function was called
        expect(onPress).toHaveBeenCalled();
    });
});
