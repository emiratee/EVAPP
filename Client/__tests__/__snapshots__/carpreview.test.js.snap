// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarPreview component renders correctly with selected car 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderColor": "#e30000",
      "borderRadius": 10,
      "borderWidth": 1,
      "height": 100,
      "justifyContent": "space-between",
      "marginRight": 10,
      "opacity": 1,
      "padding": 10,
      "width": 100,
    }
  }
  testID="car-preview-wrapper"
>
  <Text
    style={
      {
        "fontWeight": "bold",
      }
    }
  >
    Toyota Camry
  </Text>
  <Text>
    Silver
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <Text>
      5
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 18,
          },
          undefined,
          {
            "fontFamily": "material-community",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      ó±‰‰
    </Text>
  </View>
  <Text>
    ABC-123
  </Text>
</View>
`;

exports[`CarPreview component renders correctly with unselected car 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "borderColor": "#a8a8a8",
      "borderRadius": 10,
      "borderWidth": 1,
      "height": 100,
      "justifyContent": "space-between",
      "marginRight": 10,
      "opacity": 1,
      "padding": 10,
      "width": 100,
    }
  }
  testID="car-preview-wrapper"
>
  <Text
    style={
      {
        "fontWeight": "bold",
      }
    }
  >
    Toyota Camry
  </Text>
  <Text>
    Silver
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <Text>
      5
    </Text>
    <Text />
  </View>
  <Text>
    ABC-123
  </Text>
</View>
`;
